File: packages/ledger_engine/lib/src/hlc.dart
import 'dart:typed_data';

import 'package:meta/meta.dart';

/// Hybrid Logical Clock for causal ordering of events.
///
/// Format: `timestamp-counter-nodeId` where:
/// - timestamp: UTC wall-clock time
/// - counter: tiebreaker for events at the same millisecond
/// - nodeId: first 8 hex chars of the node's public key
@immutable
class HybridLogicalClock implements Comparable<HybridLogicalClock> {
  /// Creates an [HybridLogicalClock] with explicit values.
  const HybridLogicalClock({
    required this.timestamp,
    required this.counter,
    required this.nodeId,
  });

  /// Creates a new HLC for a local event.
  ///
  /// If the wall-clock has advanced past [previous], counter resets to 0.
  /// If the wall-clock is equal or behind [previous], counter increments.
  factory HybridLogicalClock.now({
    required HybridLogicalClock? previous,
    required String nodeId,
  }) {
    final now = DateTime.now().toUtc();

    if (previous == null) {
      return HybridLogicalClock(timestamp: now, counter: 0, nodeId: nodeId);
    }

    final prevMs = previous.timestamp.millisecondsSinceEpoch;
    final nowMs = now.millisecondsSinceEpoch;

    if (nowMs > prevMs) {
      return HybridLogicalClock(timestamp: now, counter: 0, nodeId: nodeId);
    }

    // Wall clock hasn't advanced: keep previous timestamp, bump counter.
    return HybridLogicalClock(
      timestamp: previous.timestamp,
      counter: previous.counter + 1,
      nodeId: nodeId,
    );
  }

  /// Updates the HLC upon receiving a remote event.
  ///
  /// Takes the max of local, remote, and wall-clock timestamps.
  factory HybridLogicalClock.receive({
    required HybridLogicalClock local,
    required HybridLogicalClock remote,
    required String nodeId,
  }) {
    final now = DateTime.now().toUtc();
    final localMs = local.timestamp.millisecondsSinceEpoch;
    final remoteMs = remote.timestamp.millisecondsSinceEpoch;
    final nowMs = now.millisecondsSinceEpoch;

    final maxMs = _max3(localMs, remoteMs, nowMs);

    int counter;
    if (maxMs == localMs && maxMs == remoteMs) {
      counter = _max2(local.counter, remote.counter) + 1;
    } else if (maxMs == localMs) {
      counter = local.counter + 1;
    } else if (maxMs == remoteMs) {
      counter = remote.counter + 1;
    } else {
      counter = 0;
    }

    return HybridLogicalClock(
      timestamp: DateTime.fromMillisecondsSinceEpoch(maxMs, isUtc: true),
      counter: counter,
      nodeId: nodeId,
    );
  }

  /// Parses a canonical HLC string.
  factory HybridLogicalClock.fromCanonical(String s) {
    // Format: ISO8601-COUNTER-NODEID
    // The ISO timestamp ends with 'Z', so split from the right.
    final lastDash = s.lastIndexOf('-');
    final nodeId = s.substring(lastDash + 1);

    final beforeNodeId = s.substring(0, lastDash);
    final secondLastDash = beforeNodeId.lastIndexOf('-');
    final counterStr = beforeNodeId.substring(secondLastDash + 1);
    final tsStr = beforeNodeId.substring(0, secondLastDash);

    return HybridLogicalClock(
      timestamp: DateTime.parse(tsStr),
      counter: int.parse(counterStr),
      nodeId: nodeId,
    );
  }

  /// UTC wall-clock time.
  final DateTime timestamp;

  /// Logical counter (tiebreaker within the same millisecond).
  final int counter;

  /// Node identifier (first 8 hex chars of pubkey).
  final String nodeId;

  /// Canonical string: `2026-02-24T12:00:00.000Z-0042-a1b2c3d4`
  String toCanonical() {
    final ts = timestamp.toIso8601String();
    final cnt = counter.toString().padLeft(4, '0');
    return '$ts-$cnt-$nodeId';
  }

  /// Serializes to bytes for inclusion in hash computations.
  Uint8List toBytes() {
    final canonical = toCanonical();
    return Uint8List.fromList(canonical.codeUnits);
  }

  @override
  int compareTo(HybridLogicalClock other) {
    final tsCmp = timestamp.compareTo(other.timestamp);
    if (tsCmp != 0) return tsCmp;

    final cntCmp = counter.compareTo(other.counter);
    if (cntCmp != 0) return cntCmp;

    return nodeId.compareTo(other.nodeId);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is HybridLogicalClock &&
          timestamp == other.timestamp &&
          counter == other.counter &&
          nodeId == other.nodeId;

  @override
  int get hashCode => Object.hash(timestamp, counter, nodeId);

  @override
  String toString() => 'HLC(${toCanonical()})';

  static int _max2(int a, int b) => a > b ? a : b;
  static int _max3(int a, int b, int c) => _max2(_max2(a, b), c);
}

File: styx-js/src/ledger/hlc.js
// styx-js/src/ledger/hlc.js
// Hybrid Logical Clock — combines wall-clock time with logical counter and node ID

import { concatBytes, utf8Encode } from '../utils.js';

/**
 * Hybrid Logical Clock combining wall-clock time, logical counter, and node ID.
 */
export class HybridLogicalClock {
  /**
   * @param {Date} timestamp - UTC wall-clock time
   * @param {number} counter - Logical counter (tiebreaker within same millisecond)
   * @param {string} nodeId - Node identifier (first 8 hex chars of pubkey)
   */
  constructor(timestamp, counter, nodeId) {
    this.timestamp = timestamp;
    this.counter = counter;
    this.nodeId = nodeId;
    Object.freeze(this);
  }

  /**
   * Create HLC for the current instant, ensuring monotonicity with previous HLC
   * @param {HybridLogicalClock|null} previous
   * @param {string} nodeId
   * @returns {HybridLogicalClock}
   */
  static now(previous, nodeId) {
    const now = new Date();

    if (!previous) {
      return new HybridLogicalClock(now, 0, nodeId);
    }

    if (now.getTime() > previous.timestamp.getTime()) {
      return new HybridLogicalClock(now, 0, nodeId);
    }

    // Wall clock hasn't advanced — increment logical counter
    return new HybridLogicalClock(previous.timestamp, previous.counter + 1, nodeId);
  }

  /**
   * Parse from canonical format: "2026-02-24T12:00:00.000Z-0042-a1b2c3d4"
   * @param {string} s
   * @returns {HybridLogicalClock}
   */
  static fromCanonical(s) {
    const parts = s.split('-');
    // Rejoin the ISO date parts (which contain dashes themselves)
    // Format: YYYY-MM-DDTHH:MM:SS.sssZ-CCCC-NNNNNNNN
    // The timestamp portion is everything before the counter, which is a 4-digit hex
    // Strategy: find the Z, everything before Z+1 is timestamp, then -counter-nodeId
    const zIdx = s.indexOf('Z');
    if (zIdx === -1) throw new Error(`Invalid HLC canonical format: ${s}`);

    const timestampStr = s.substring(0, zIdx + 1);
    const rest = s.substring(zIdx + 2); // skip "Z-"
    const dashIdx = rest.indexOf('-');
    if (dashIdx === -1) throw new Error(`Invalid HLC canonical format: ${s}`);

    const counterStr = rest.substring(0, dashIdx);
    const nodeId = rest.substring(dashIdx + 1);

    return new HybridLogicalClock(
      new Date(timestampStr),
      parseInt(counterStr, 10),
      nodeId
    );
  }

  /**
   * Canonical string: "2026-02-24T12:00:00.000Z-0042-a1b2c3d4"
   * @returns {string}
   */
  toCanonical() {
    const ts = this.timestamp.toISOString();
    const counter = this.counter.toString().padStart(4, '0');
    return `${ts}-${counter}-${this.nodeId}`;
  }

  /**
   * Serialize to bytes for hash computation
   * @returns {Uint8Array}
   */
  toBytes() {
    return utf8Encode(this.toCanonical());
  }

  /**
   * Compare by timestamp, then counter, then nodeId
   * @param {HybridLogicalClock} other
   * @returns {number} -1, 0, or 1
   */
  compareTo(other) {
    const tsDiff = this.timestamp.getTime() - other.timestamp.getTime();
    if (tsDiff !== 0) return tsDiff < 0 ? -1 : 1;
    if (this.counter !== other.counter) return this.counter < other.counter ? -1 : 1;
    if (this.nodeId < other.nodeId) return -1;
    if (this.nodeId > other.nodeId) return 1;
    return 0;
  }

  toString() {
    return this.toCanonical();
  }
}
